diff --git a/src/planner/expression_binder/base_select_binder.cpp b/src/planner/expression_binder/base_select_binder.cpp
index e0adf5398d..d0c8cd758a 100644
--- a/src/planner/expression_binder/base_select_binder.cpp
+++ b/src/planner/expression_binder/base_select_binder.cpp
@@ -101,9 +101,9 @@ BindResult BaseSelectBinder::BindGroup(ParsedExpression &expr, idx_t depth, idx_
 	if (it != info.collated_groups.end()) {
 		// This is an implicitly collated group, so we need to refer to the first() aggregate
 		const auto &aggr_index = it->second;
-		auto uncollated_first_expression =
-		    make_uniq<BoundColumnRefExpression>(expr.GetName(), node.aggregates[aggr_index]->return_type,
-		                                        ColumnBinding(node.aggregate_index, aggr_index), depth);
+		const auto return_type = node.aggregates[aggr_index]->return_type;
+		auto uncollated_first_expression = make_uniq<BoundColumnRefExpression>(
+		    expr.GetName(), return_type, ColumnBinding(node.aggregate_index, aggr_index), depth);
 
 		if (node.groups.grouping_sets.size() <= 1) {
 			// if there are no more than two grouping sets, you can return the uncollated first expression.
@@ -117,10 +117,10 @@ BindResult BaseSelectBinder::BindGroup(ParsedExpression &expr, idx_t depth, idx_
 		auto collated_group_expression = make_uniq<BoundColumnRefExpression>(
 		    expr.GetName(), group->return_type, ColumnBinding(node.group_index, group_index), depth);
 
-		auto sql_null = make_uniq<BoundConstantExpression>(Value(LogicalType::VARCHAR));
+		auto sql_null = make_uniq<BoundConstantExpression>(Value(return_type));
 		auto when_expr = make_uniq<BoundOperatorExpression>(ExpressionType::OPERATOR_IS_NULL, LogicalType::BOOLEAN);
 		when_expr->children.push_back(std::move(collated_group_expression));
-		auto then_expr = make_uniq<BoundConstantExpression>(Value(LogicalType::VARCHAR));
+		auto then_expr = make_uniq<BoundConstantExpression>(Value(return_type));
 		auto else_expr = std::move(uncollated_first_expression);
 		auto case_expr =
 		    make_uniq<BoundCaseExpression>(std::move(when_expr), std::move(then_expr), std::move(else_expr));
