diff --git a/src/common/arrow/arrow_converter.cpp b/src/common/arrow/arrow_converter.cpp
index 5369b3d423..2348726934 100644
--- a/src/common/arrow/arrow_converter.cpp
+++ b/src/common/arrow/arrow_converter.cpp
@@ -116,11 +116,15 @@ void SetArrowFormat(DuckDBArrowSchemaHolder &root_holder, ArrowSchema &child, co
                     ClientProperties &options, ClientContext &context) {
 	if (type.HasAlias()) {
 		// If it is a json type, we only export it as json if arrow_lossless_conversion = True
-		if (!(type.IsJSONType() && !options.arrow_lossless_conversion)) {
-			// If the type has an alias, we check if it is an Arrow-Type extension
-			if (SetArrowExtension(root_holder, child, type, context)) {
-				return;
-			}
+		// if (!(type.IsJSONType() && !options.arrow_lossless_conversion)) {
+		// 	// If the type has an alias, we check if it is an Arrow-Type extension
+		// 	if (SetArrowExtension(root_holder, child, type, context)) {
+		// 		return;
+		// 	}
+		// }
+		// If the type has an alias, we check if it is an Arrow-Type extension
+		if (SetArrowExtension(root_holder, child, type, context)) {
+			return;
 		}
 	}
 	switch (type.id()) {
diff --git a/src/common/arrow/arrow_type_extension.cpp b/src/common/arrow/arrow_type_extension.cpp
index dee514cad4..67edc6abd2 100644
--- a/src/common/arrow/arrow_type_extension.cpp
+++ b/src/common/arrow/arrow_type_extension.cpp
@@ -380,7 +380,7 @@ void ArrowTypeExtensionSet::Initialize(const DBConfig &config) {
 
 	// Types that are 1:n
 	config.RegisterArrowExtension({"arrow.json", &ArrowJson::PopulateSchema, &ArrowJson::GetType,
-	                               make_shared_ptr<ArrowTypeExtensionData>(LogicalType::VARCHAR)});
+	                               make_shared_ptr<ArrowTypeExtensionData>(LogicalType::JSON())});
 
 	config.RegisterArrowExtension({"DuckDB", "bit", &ArrowBit::PopulateSchema, &ArrowBit::GetType,
 	                               make_shared_ptr<ArrowTypeExtensionData>(LogicalType::BIT), nullptr, nullptr});
